src/coprocessor/endpoint.rs

```diff
@@ -291,39 +266,46 @@ impl<E: Engine> Endpoint<E> {
             })
     }
 
-    /// Handle a unary request and run on the read pool. Returns a future producing the
-    /// result, which must be a `Response` and will never fail. If there are errors during
-    /// handling, they will be embedded in the `Response`.
+    /// Handle a unary request and run on the read pool.
+    ///
+    /// Returns `Err(err)` if the read pool is full. Returns `Ok(future)` in other cases.
+    /// The future inside may be an error however.
     fn handle_unary_request(
         &self,
         req_ctx: ReqContext,
         handler_builder: RequestHandlerBuilder<E::Snap>,
-    ) -> impl Future<Item = coppb::Response, Error = ()> {
+    ) -> Result<impl Future<Item = coppb::Response, Error = Error>> {
         let engine = self.engine.clone();
         let priority = readpool::Priority::from(req_ctx.context.get_priority());
+        // box the tracker so that moving it is cheap.
         let mut tracker = box Tracker::new(req_ctx);
 
-        let result = self.read_pool.future_execute(priority, move |ctxd| {
-            tracker.attach_ctxd(ctxd);
-
-            Self::handle_unary_request_impl(engine, tracker, handler_builder)
-        });
-
-        future::result(result)
-            // If the read pool is full, an error response will be returned directly.
+        self.read_pool
+            .future_execute(priority, move |ctxd| {
+                tracker.attach_ctxd(ctxd);
+                Self::handle_unary_request_impl(engine, tracker, handler_builder)
+            })
             .map_err(|_| Error::Full)
```

Comment:
I don't think it's correct anymore. What if the returned error is a `Error::Region`?

Reply:
@BusyJay `future_execute` itself returns `Result<impl Future>`. The `map_err` here only maps the out most `Result`, not the inner one. For the out most result, it can be only caused by read pool full. Does it solve your concern?

Reply:
Why not `impl From` for the error returned by `read_pool`? So no one has the concern anymore.

Reply:
This solution was not accepted at that time.